### PR TITLE
feat(query): agg-hashtable-p2

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

Introduce agg hashtable in the query pipeline and support partition ht.

We can enable this via : `set enable_experimental_aggregate_hashtable = 1;`

Note this pr only works in singleton deployment currently.

Pros:
1. better performance in high cardinality group aggregation, up to 2x
2. better and cleaner codes than the old one

Cons:
1. It may consume more memory usage than the old one cause it uses using two-part struct layout
2. A little worse performance in normal group aggregation,  ~ 0.9x

- Closes #issue

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13548)

### CODE REVIEW

**Comment by sundy-li** [comment]
Performance tests:

Queries:
```
select (number % 200)  as a  from numbers(1000000000) group by all order by a desc limit 3;

select (number % 200)  as a, (number % 300)  as b   from numbers(1000000000) group by all order by a desc limit 3;

select (number % 200)  as a, (number % 300)  as b, (number % 400)  as c from numbers(1000000000) group by all order by a desc limit 3;

select (number % 200)  as a, (number % 300)  as b, (number % 400)  as c, (number % 500)  as d   from numbers(1000000000) group by all order by a desc limit 3;

select (number % 200)  as a, (number % 300)  as b, (number % 400)  as c, (number % 500)  as d, (number % 600)  as e  from numbers(1000000000) group by all order by a desc limit 3;


```

No improvement, it's unexpected.

image

File: src/query/expression/src/kernels/utils.rs [hunk 1]
```diff
@@ -68,17 +68,8 @@ pub unsafe fn set_vec_len_by_ptr<T>(vec: &mut Vec<T>, ptr: *const T) {
 /// # Safety
 /// # As: copy_nonoverlapping
```
Comment (RinChanNOWWW) [inline src/query/expression/src/kernels/utils.rs:69]: The comment should be updated.

**Comment by BohuTANG** [comment]
TPCH-SF100 native test:

```
set global enable_experimental_aggregate_hashtable=1;
```

## Result

Query  | xsmall-new-hashtable | xsmall-raw-hashtable
-- | -- | --
TPC-H 1 | 10.514 | 10.631
TPC-H 2 | 5.212 | 4.333
TPC-H 3 | 9.669 | 8.251
TPC-H 4 | 5.42 | 5.3
TPC-H 5 | 10.598 | 10.085
TPC-H 6 | 4.085 | 4.026
TPC-H 7 | 10.681 | 10.266
TPC-H 8 | 13.597 | 13.583
TPC-H 9 | 21.921 | 21.711
TPC-H 10 | 17.692 | 16.442
TPC-H 11 | 2.622 | 2.594
TPC-H 12 | 9.149 | 8.911
TPC-H 13 | 11.665 | 12.603
TPC-H 14 | 6.299 | 6.29
TPC-H 15 | 6.424 | 6.08
TPC-H 16 | 4.175 | 1.796
TPC-H 17 | OOM | 27.24
TPC-H 18 | OOM | 34.936
TPC-H 19 | OOM | 14.342
TPC-H 20 | OOM | 12.327
TPC-H 21 | 23.468 | 20.042
TPC-H 22 | 2.661 | 2.476

# How to test

```
export BENDSQL_DSN=<>
bendsql> create warehouse 'bh-pr13548' warehouse_size='xsmall' with version='pr-13548-308de57' cache_size=0;
```

https://github.com/BohuTANG/wizard/tree/main/benchsb/bend

```
export BENDSQL_DSN=<>
python3 bend.py --nosuspend
```

**Comment by sundy-li** [comment]
hits benchmark:

image

Q31, Q32
image

**Pull request closed without merging**